src/parser.rs

```diff
@@ -812,22 +828,32 @@ fn parse_word_parts(
     ))(input)?;
 
     let mut result = Vec::new();
-    for part in parts {
+    let mut parts = parts.into_iter().enumerate().peekable();
+    while let Some((i, part)) = parts.next() {
       match part {
         PendingPart::Char(c) => {
-          if let Some(WordPart::Text(text)) = result.last_mut() {
-            text.push(c);
+          append_char(&mut result, c);
+        }
+        PendingPart::Tilde => {
+          if i == 0 {
+            if matches!(parts.peek(), None | Some((_, PendingPart::Char('/'))))
+            {
+              result.push(WordPart::Tilde);
+            } else {
+              return ParseError::fail(
+                original_input,
+                "Unsupported tilde expansion.",
+              );
```

Comment:
There's lots of different kinds of tilde expansion. I don't think they're all worth supporting: https://www.gnu.org/software/bash/manual/html_node/Tilde-Expansion.html